crates/circuit/src/classical/expr/range.rs

```
@@ -0,0 +1,331 @@
+                        }
+                    };
+
+                    (start_expr, stop_expr, step_expr)
+                }
+                _ => {
+                    // If not a Uint type, we can't cast
+                    return Err(PyErr::new::<pyo3::exceptions::PyTypeError, _>(
+                        "Range type must be an unsigned integer type",
+                    ));
+                }
+            }
+        };
+
+        Ok((
+            PyRange(Range {
+                start: start_expr,
+                stop: stop_expr,
+                step: Some(step_expr),
+                ty: target_ty,
+                constant,
+            }),
+            PyExpr(ExprKind::Range),
+        ))
+    }
+
+    #[getter]
+    fn get_start(&self, py: Python) -> PyResult<Py<PyAny>> {
+        self.0.start.clone().into_py_any(py)
+    }
+
+    #[getter]
+    fn get_stop(&self, py: Python) -> PyResult<Py<PyAny>> {
+        self.0.stop.clone().into_py_any(py)
+    }
+
+    #[getter]
+    fn get_step(&self, py: Python) -> PyResult<Py<PyAny>> {
+        self.0.step.as_ref().unwrap().clone().into_py_any(py)
+    }
+
+    #[getter]
+    fn get_const(&self) -> bool {
+        self.0.constant
+    }
+
+    #[getter]
+    fn get_type(&self, py: Python) -> PyResult<Py<PyAny>> {
+        self.0.ty.into_py_any(py)
+    }
+
+    fn __len__(&self) -> usize {
+        1
+    }
+
+    fn accept<'py>(
+        slf: PyRef<'py, Self>,
+        visitor: &Bound<'py, PyAny>,
+    ) -> PyResult<Bound<'py, PyAny>> {
+        visitor.call_method1(intern!(visitor.py(), "visit_range"), (slf,))
+    }
+
+    fn __repr__(&self, py: Python) -> PyResult<String> {
+        let start = self.0.start.clone().into_py_any(py)?.bind(py).repr()?;
+        let stop = self.0.stop.clone().into_py_any(py)?.bind(py).repr()?;
+        let step = format!(
+            ", step={}",
+            self.0
+                .step
+                .as_ref()
+                .unwrap()
```

`unwrap` will panic when you can return an exception, but more than that - will this fail simply by `step` being `None`? This means trying to print a range with `None` step will crash the program. If this is the case we should handle this differently.